File: src/main/java/ir/reservs/reservs/data/network/ApiHelper.java
package ir.reservs.reservs.data.network;

import java.util.List;

import io.reactivex.Single;
import ir.reservs.reservs.model.ChangePassword;
import ir.reservs.reservs.model.ReserveHistory;
import ir.reservs.reservs.model.Salon;
import ir.reservs.reservs.model.Success;
import ir.reservs.reservs.model.Time;
import ir.reservs.reservs.model.User;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface ApiHelper {
    @POST("auth/register")
    @FormUrlEncoded
    Single<User> register(@Field("name") String name,
                          @Field("phone_number") String phone,
                          @Field("password") String password);


    @POST("auth/login")
    @FormUrlEncoded
    Single<User> login(@Field("phone") String phone, @Field("password") String password);

    @GET("reserves")
    Single<List<ReserveHistory>> reserves();

    @POST("reserve")
    @FormUrlEncoded
    Single<Success> reserve(@Field("time_id") String time_id,
                            @Field("salon_id") Integer salon_id,
                            @Field("authority") String authority,
                            @Field("date") String date);

    @POST("reserve/update")
    @FormUrlEncoded
    Single<Success> reserveUpdate(@Field("authority") String authority);

    @GET("salons")
    Single<List<Salon>> salons();

    @PUT("info")
    @FormUrlEncoded
    Single<Success> updateName(@Field("name") String name);


    @PUT("password")
    @FormUrlEncoded
    Single<ChangePassword> updatePassword(@Field("current_password") String current_password,
                                          @Field("new_password") String new_password);

    @GET("times/{salon_id}/{date}")
    Single<List<Time>> times(@Path("salon_id") int salon_id, @Path("date") String date);

}

File: src/main/java/ir/reservs/reservs/data/network/ApiHelper.kt
package ir.reservs.reservs.data.network

import io.reactivex.Single
import ir.reservs.reservs.model.*
import okhttp3.MultipartBody
import retrofit2.http.*

interface ApiHelper {
    @POST("auth/register")
    @FormUrlEncoded
    fun register(@Field("name") name: String,
                 @Field("phone_number") phone: String,
                 @Field("password") password: String,
                 @Field("fcmToken") fcmToken: String): Single<User>


    @POST("auth/login")
    @FormUrlEncoded
    fun login(@Field("phone") phone: String,
              @Field("password") password: String,
              @Field("fcmToken") fcmToken: String): Single<User>

    @POST("auth/phone")
    @FormUrlEncoded
    fun sendPhone(@Field("phone") phone: String): Single<Message>

    @POST("auth/confirm")
    @FormUrlEncoded
    fun confirmCode(@Field("phone") phone: String, @Field("code") code: String): Single<User>

    @POST("auth/forget")
    @FormUrlEncoded
    fun forget(@Field("phone_number") phone: String): Single<Success>

    @GET("reserves")
    fun reserves(@Query("page") page: Int): Single<MutableList<History>>

    @POST("reserve")
    @FormUrlEncoded
    fun reserve(@Field("time_id") time_id: String,
                @Field("salon_id") salon_id: Int?,
                @Field("CallbackURL") callBackUrl: String,
                @Field("date") date: String): Single<Payment>

    @POST("reserve/update")
    @FormUrlEncoded
    fun reserveUpdate(@Field("authority") authority: String): Single<Success>

    @GET("salons")
    fun salons(): Single<MutableList<Salon>>

    @GET("city/{city}/salons")
    fun salons(@Path("city") cityId: Int): Single<MutableList<Salon>>

    @PUT("info")
    @FormUrlEncoded
    fun updateName(@Field("name") name: String): Single<Success>

    @PUT("password")
    @FormUrlEncoded
    fun updatePassword(@Field("current_password") current_password: String,
                       @Field("new_password") new_password: String): Single<ChangePassword>

    @GET("times/{salon_id}/{date}")
    fun times(@Path("salon_id") salon_id: Int, @Path("date") date: String): Single<MutableList<Time>>

    @PUT("updateFcmToken")
    @FormUrlEncoded
    fun updateFcmToken(@Field("fcmToken") fcmToken: String): Single<Success>

    @GET("cities")
    fun cities(): Single<MutableList<City>>

    @Multipart
    @POST("avatar")
    fun updateAvatar(@Part avatar: MultipartBody.Part): Single<Avatar>
}